### PR TITLE
Added maxRows and filter support

### PR DESCRIPTION
**Pull request by gwynjudd**
Useful for the display of very large swagger specs. Can limit the number of operations to a smaller value and search.

For example:

  ```
const ui = SwaggerUIBundle({
    url: "http://petstore.swagger.io/v2/swagger.json",
    dom_id: '#swagger-ui',
    docExpansion: 'none',
    maxRows: 100,
    filter: '',
    presets: [
      SwaggerUIBundle.presets.apis,
      SwaggerUIStandalonePreset
    ],
    plugins: [
      SwaggerUIBundle.plugins.DownloadUrl
    ],
    layout: "StandaloneLayout"
  })
```

Appears as:

image

We can filter the list:

image

### CODE REVIEW

**Comment by gwynjudd** [comment]
Helps to resolve #1919 for me

**Comment by webron** [comment]
Interesting. Please allow me some time to play with it and evaluate these changes (functionally). 
Regarding the filter, there have been similar requests in the past. It might be a starting point to an expanded functionality (like filtering based on operation verb and so on).

What is the filter based on now? The path name? Description?

**Comment by gwynjudd** [comment]
@webron it just filters tagged operations based on the tag name. There are obvious ways it could be improved but I thought this was a simple start to show the approach

**Comment by webron** [comment]
@gwynjudd just to give you a heads up - not sure I'll finish reviewing the feature before today's release, but I'll make sure to give you feedback in time for this to make it into next week's release (if makes sense, of course).

**Comment by webron** [comment]
Waiting for @shockey's feedback as well, but here are a few thoughts:

- It feels like this should have been in two separate PRs, one for each new config. They are not directly connected.
- It took me a bit to understand that the value given to `filter` is the string that's going to be used. I thought it was going to be `true`/`false` to enabled/disable. I'm ok with the current behavior, but would expect the boolean values of `true`/`false` to not be treated as strings. `true` should be treated as enabling (with the value of '') and `false` should be an explicit disable. That would lead to a more predictable behavior.
- The design of the `filter` rendering should be reviewed by @Minasokoni. It feels like it's in the wrong place, perhaps needs better styling.
- `maxRows` doesn't do what I expected it to. It limits the number of *tags* and not the number of *operations*. Would suggest changing the config name to `maxDisplayedTags` or `maxRenderedTags`.
- `maxRows` values are a bit unpredictable if you use non-numeric values such as boolean, non-numeric string. Would suggest that in all those cases, the value is ignored rather than the current behavior. Consider applying the same to sub-zero values.

**Comment by webron** [comment]
Sounds good, @Minasokoni. @gwynjudd - I assume you wouldn't mind that but let us know otherwise.

**Comment by gwynjudd** [comment]
@webron @Minasokoni sorry I've been a bit quiet this week, I've been totally snowed under at work. I agree with all of your suggestions, it all sounds fine to me

**Comment by gwynjudd** [comment]
@webron I probably won't have time this week to make any changes, but I can probably do something next week if you want me to make any changes

**Comment by webron** [comment]
@gwynjudd sure, there's no hurry. We'll wait for an update from you.

**Comment by shockey** [comment]
@gwynjudd, nudge 😄

**Comment by gwynjudd** [comment]
@shockey hello, sorry I haven't really had any time right now to work on it. Work is super busy right now but when I get some down-time I'll get it cleaned up.

**Comment by Minasokoni** [comment]
@shockey my bad too. I haven't checked this out yet. I can tomorrow though

**Comment by gwynjudd** [comment]
@webron I wasn't sure when I was going to have time to look at this, but then I figured I could at least make some small changes today rather than just say I didn't know when I would be able to do anything on it. Currently work is in a pretty busy phase so it's hard to find the time.

Anyway I made the following changes:

* changed `maxRows` to `maxDisplayedTags`
* changed `filter` to allow true/false as well as explicit filter strings
* changed `maxDisplayedTags` to ignore non numeric values as well as < 0
* updated readme

> It feels like this should have been in two separate PRs, one for each new config. They are not directly connected.

To me the features are very closely related. For instance, if you have got a `maxDisplayedTags` set, then it becomes necessary to also have a filter set, otherwise it will be impossible to see the tags that are not displayed. Please let me know if you still want me to split them out, and I will try to make that effort.

> The design of the filter rendering should be reviewed by @Minasokoni. It feels like it's in the wrong place, perhaps needs better styling.

@Minasokoni if you can, please do so, front end design isn't my strong point

**Comment by webron** [comment]
Thanks for the modifications, @gwynjudd. Normally, I'd prefer two PRs to make the review process easier and remove dependencies, but in this case it's ok.

Just need @Minasokoni's input.

File: src/core/components/operations.jsx [hunk 1]
```diff
@@ -33,7 +35,17 @@ export default class Operations extends React.Component {
     const Collapse = getComponent("Collapse")
 
     let showSummary = layoutSelectors.showSummary()
-    let { docExpansion, displayOperationId, displayRequestDuration } = getConfigs()
+    let { docExpansion, displayOperationId, displayRequestDuration, maxDisplayedTags } = getConfigs()
+
+    if (filter) {
+      taggedOps = taggedOps.filter((tagObj, tag) => {
+        return tag.startsWith(filter === true ? "" : filter)
```
Comment (heldersepu) [inline src/core/components/operations.jsx]: Why limit to startsWith ?
I think it would be better to use indexOf to find any matches.

**Comment by gwynjudd** [comment]
@webron I'm not sure what happened just now, I thought I was bringing my branch into sync with master, but something odd happened and there are a great many unexpected commits that I didn't make. Do you have any suggestions on how to fix that? Sorry about this

File: src/core/components/operations.jsx (hunk 1 above)
Comment (gwynjudd) [inline src/core/components/operations.jsx]: It should be fixed with my latest change, but I think I also made an error syncing up with master so the PR is a bit confused at this time

**Comment by gwynjudd** [comment]
OK I fixed my mistake from yesterday, it now contains only my expected commits

**Comment by shockey** [comment]
Merged. 

Thanks for the feature @gwynjudd, and thanks for bearing with us through the review process :smile:

**Comment by heldersepu** [comment]
Awesome, I'm looking forward to incorporate this feature in my project! 
...any chance we can get a `Rebuild dist` soon?

**Comment by shockey** [comment]
@heldersepu, there will be one along with the new release tonight 😄

**Comment by Aarbel** [comment]
@gwynjudd, the implementation of Swagger plugins isn't really clear for me, how could i add the filter feature in my code ? 

Thanks a lot ! 

```
const path = require('path')
const replace = require('replace');
const express = require('express')
const app = express()

const swaggerUiAssetPath = require("swagger-ui-dist").absolutePath()

const { LOCALHOST_URL } = process.env

replace({
    regex: 'http.*swagger.json',
    replacement : LOCALHOST_URL + ':3012/public/swagger.json',
    paths: [ path.join(swaggerUiAssetPath, 'index.html') ],
    recursive:false,
    silent:true,
})

app
    .use('/', express.static(swaggerUiAssetPath))
    .use('/public',express.static('/usr/src/documentation/public'))

    .listen(3000)
```

**Comment by gwynjudd** [comment]
@Aarbel not too sure sorry, it would have to be where you are configuring the swagger UI bundle itself, but I cant see that in your code example. Maybe it is in your `swagger-ui-dist` location? Somewhere there is a line kind of like this:

```
const ui = SwaggerUIBundle({
...
});
```

In the `...` bit, you would be passing some options, and you would configure the `filter` option there and `maxDisplayedTags` there.

**Comment by shockey** [comment]
@Aarbel, you'd need to modify the `index.html` file that comes with Swagger-UI's dist package and serve that instead of the stock one. From within that file, you can provide custom plugins and modify the markup around Swagger-UI.

**Comment by Aarbel** [comment]
@shockey @gwynjudd thanks for your help ;) the `swagger-ui-dist` location is directly in my node_modules folder, 
cf my package.json : 
```
"devDependencies": {
    "replace": "^0.3.0",
    "swagger-ui-dist": "^3.0.18"
  }
```
I'm not a big fan about packages modification, is there an other way do add plugins & SwaggerUIBundle's options without editing it ?

**Comment by shockey** [comment]
@Aarbel, not really, you need to serve your own HTML that invokes Swagger-UI in order to customize it. I'd suggest just crafting your own `index.html` that calls Swagger-UI how you'd like to, and serving that instead of the `index.html` that comes with the module.

I know it feels a bit weird, but the `index.html` asset isn't _really_ part of Swagger-UI itself, it's more of a convenience for folks that just want to get up and running with little effort 😄

**Pull request merged**